src/Processors/QueryPlan/Optimizations/filterPushDown.cpp

```diff
@@ -204,6 +209,230 @@ static size_t simplePushDownOverStep(QueryPlan::Node * parent_node, QueryPlan::N
+
+        const auto & input_header = push_to_left_stream ? left_stream_input_header : right_stream_input_header;
+        const auto & input_columns_names = input_header.getNames();
+
+        for (const auto & name : input_columns_names)
+        {
+            if (!join_header.has(name))
+                continue;
+
+            /// Skip if type is changed. Push down expression expect equal types.
+            if (!input_header.getByName(name).type->equals(*join_header.getByName(name).type))
+                continue;
+
+            available_input_columns_for_filter.push_back(name);
+        }
+
+        return available_input_columns_for_filter;
+    };
+
+    bool left_stream_filter_push_down_input_columns_available = true;
+    bool right_stream_filter_push_down_input_columns_available = true;
+
+    if (table_join.kind() == JoinKind::Left)
+        right_stream_filter_push_down_input_columns_available = false;
+    else if (table_join.kind() == JoinKind::Right)
+        left_stream_filter_push_down_input_columns_available = false;
+
+    /** We disable push down to right table in cases:
+      * 1. Right side is already filled. Example: JOIN with Dictionary.
+      * 2. ASOF Right join is not supported.
+      */
+    bool allow_push_down_to_right = join && join->allowPushDownToRight() && table_join.strictness() != JoinStrictness::Asof;
+    if (!allow_push_down_to_right)
+        right_stream_filter_push_down_input_columns_available = false;
+
+    Names equivalent_columns_to_push_down;
+
+    if (left_stream_filter_push_down_input_columns_available)
+    {
+        for (const auto & [name, _] : equivalent_left_stream_column_to_right_stream_column)
+            equivalent_columns_to_push_down.push_back(name);
+    }
+
+    if (right_stream_filter_push_down_input_columns_available)
+    {
+        for (const auto & [name, _] : equivalent_right_stream_column_to_left_stream_column)
+            equivalent_columns_to_push_down.push_back(name);
+    }
+
+    Names left_stream_available_columns_to_push_down = get_available_columns_for_filter(true /*push_to_left_stream*/, left_stream_filter_push_down_input_columns_available);
+    Names right_stream_available_columns_to_push_down = get_available_columns_for_filter(false /*push_to_left_stream*/, right_stream_filter_push_down_input_columns_available);
+
+    auto join_filter_push_down_actions = filter->getExpression()->splitActionsForJOINFilterPushDown(filter->getFilterColumnName(),
+        filter->removesFilterColumn(),
+        left_stream_available_columns_to_push_down,
+        left_stream_input_header.getColumnsWithTypeAndName(),
+        right_stream_available_columns_to_push_down,
+        right_stream_input_header.getColumnsWithTypeAndName(),
+        equivalent_columns_to_push_down,
+        equivalent_left_stream_column_to_right_stream_column,
+        equivalent_right_stream_column_to_left_stream_column);
+
+    size_t updated_steps = 0;
```

Comment:
I remember there was an issue with the assumption that all optimization of that kind should be idempotent, because we apply them until it can't be applied. How do we handle this case? For example, when we changed `t1.x > 0` to `t0.x > 0` and pushed it into the left, on the second attempt we still have `t1.x > 0` in WHERE, and we can push it once more. Where is termination checked?

Reply:
Yes, if we have `t1.x > 0` and for example it is INNER JOIN, we push `t1.x > 0` to LEFT side, `t2.x > 0` to RIGHT side, and we remove `t1.x > 0` completely from WHERE in `splitActionsForJOINFilterPushDown`.
After `ActionsDAG::splitActionsForJOINFilterPushDown` call, we get DAG that we need to push to JOIN LEFT side, DAG that we need to push to JOIN RIGHT side and original DAG is updated aswell to contain only rejected conditions.